sys/net/routing/olsr2/node.c

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright (C) 2014 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser General
+ * Public License v2.1. See the file LICENSE in the top level directory for more
+ * details.
+ *
+ * @ingroup olsr2
+ * @{
+ * @author  Benjamin Valentin <benpicco@zedat.fu-berlin.de>
+ * @}
+ */
+
+#include <stdio.h>
+
+#include "node.h"
+#include <slist.h>
+#include "debug.h"
+
+#include "common/netaddr.h"
+#include "rfc5444/rfc5444.h"
+
+static struct netaddr_rc local_addr;
+static struct avl_tree olsr_head;
+
+#ifdef ENABLE_DEBUG
+static struct netaddr_str nbuf[2];
+#endif
+
+#ifdef ENABLE_NAME
+char *olsr2_local_name;
+#endif
+
+static void _decrease_mpr_neigh(struct olsr_node *node)
+{
+    DEBUGF("%s (%s)", netaddr_to_str_s(&nbuf[0], node->addr), node->name);
+
+    /* only consider 2-hop nieghbors (only 2-hop neighbors have flood_mpr set) */
+    if (node->flood_mpr == NULL) {
+        return;
+    }
+
+    /* update routing MPR information */
+    struct nhdp_node *n1 = h1_deriv(get_node(node->next_addr));
+
+    if (n1 != NULL && n1->mpr_neigh_route > 0) {
+        n1->mpr_neigh_route--;
+    }
+
+    /* update flooding MPR information */
+    struct nhdp_node *n1_f = h1_deriv(get_node(node->flood_mpr));
+
+    if (n1_f != NULL && n1_f->mpr_neigh_flood > 0) {
+        n1_f->mpr_neigh_flood--;
+    }
+}
+
+static int _addr_cmp(const void *a, const void *b)
+{
+    return memcmp(a, b, NETADDR_MAX_LENGTH);
+}
+
+int olsr_node_cmp(struct olsr_node *a, struct olsr_node *b)
+{
+    return netaddr_cmp(a->addr, b->addr);
+}
+
+void node_init(void)
+{
+    local_addr._refs = 1;
+    avl_init(get_olsr_head(), _addr_cmp, false);
+}
+
+struct netaddr *get_local_addr(void)
```

Comment:
might make sense to inline this function or keep a global pointer already casted to &local_addr. This function gets called quite frequently.